src/stcal/ramp_fitting/ols_cas22/_fit_ramps.pyx

```diff
@@ -0,0 +1,138 @@
+    variances: np.ndarray
+    dq: np.ndarray
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def fit_ramps(np.ndarray[float, ndim=2] resultants,
+              np.ndarray[int, ndim=2] dq,
+              np.ndarray[float, ndim=1] read_noise,
+              float read_time,
+              list[list[int]] read_pattern,
+              bool use_jump=False,
+              float intercept=DefaultIntercept,
+              float constant=DefaultConstant):
+    """Fit ramps using the Casertano+22 algorithm.
+        This implementation uses the Cas22 algorithm to fit ramps, where
+        ramps are fit between bad resultants marked by dq flags for each pixel
+        which are not equal to zero. If use_jump is True, it additionally uses
+        jump detection to mark additional resultants for each pixel as bad if
+        a jump is suspected in them.
+
+    Parameters
+    ----------
+    resultants : np.ndarry[n_resultants, n_pixel]
+        the resultants in electrons
+    dq : np.ndarry[n_resultants, n_pixel]
+        the dq array.  dq != 0 implies bad pixel / CR.
+    read_noise : np.ndarray[n_pixel]
+        the read noise in electrons for each pixel
+    read_time : float
+        Time to perform a readout. For Roman data, this is FRAME_TIME.
+    read_pattern : list[list[int]]
+        the read pattern for the image
+    use_jump : bool
+        If True, use the jump detection algorithm to identify CRs.
+        If False, use the DQ array to identify CRs.
+    intercept : float
+        The intercept value for the threshold function. Default=5.5
+    constant : float
+        The constant value for the threshold function. Default=1/3.0
+
+    Returns
+    -------
+    A RampFitOutputs tuple
+    """
+    cdef int n_pixels, n_resultants
+    n_resultants = resultants.shape[0]
+    n_pixels = resultants.shape[1]
+
+    if n_resultants != len(read_pattern):
+        raise RuntimeError(f'The read pattern length {len(read_pattern)} does not '
+                           f'match number of resultants {n_resultants}')
+
+    # Pre-compute data for all pixels
+    cdef FixedValues fixed = fixed_values_from_metadata(metadata_from_read_pattern(read_pattern, read_time),
+                                                        Thresh(intercept, constant),
+                                                        use_jump)
+
+    # Compute all the initial sets of ramps
+    cdef deque[stack[RampIndex]] pixel_ramps = init_ramps(dq)
+
+    # Use list because this might grow very large which would require constant
+    #    reallocation. We don't need random access, and this gets cast to a python
+    #    list in the end.
+    cdef cpp_list[RampFits] ramp_fits
+
+    cdef np.ndarray[float, ndim=2] parameters = np.zeros((n_pixels, 2), dtype=np.float32)
+    cdef np.ndarray[float, ndim=2] variances = np.zeros((n_pixels, 3), dtype=np.float32)
+
+    # Copy the dq array so we can modify it without fear
+    cdef np.ndarray[int, ndim=2] fit_dq = dq.copy()
```

Comment:
Would it make sense to modify dq in place?  I have been trying to avoid making copies of any of the things that are n_pixel * n_resultant in size.